annotations_ingester/annotations_generator.py

```diff
@@ -0,0 +1,82 @@
+import logging
+import re
+from typing import Sequence
+
+from eodhp_utils.messagers import CatalogueChangeBodyMessager, Messager
+from rdflib import Graph
+
+
+class AnnotationsMessager(CatalogueChangeBodyMessager):
+    """
+    Generates basic DCAT for catalogue entries. Supports Catalogs and Collections and is
+    intended only to be sufficient for finding QA information linked to a dataset.
+
+    The output is sent to the catalogue public static files bucket under a path matching the
+    catalogue API endpoint for the dataset. For example:
+        /catalogue/
+    """
+
+    def process_delete(self, **kwargs):
+        return []
+
+    def process_update_body(
+        self,
+        entry_body: dict,
+        cat_path: str,
+        source: str,
+        target: str,
+        **kwargs,
+    ) -> Sequence[Messager.Action]:
+
+        short_path = "/".join(cat_path.split("/")[:-1])
+
+        graph = Graph()
+        graph.parse(entry_body, format="trig")
+
+        uuid = get_uuid_from_graph(entry_body.decode("utf-8"))
+
+        if uuid:
+            cache_control = 60 * 60 * 24 * 7  # 1 week
+        else:
+            cache_control = 0
+
+        turtle = graph.serialize(format="turtle")
+        jsonld = graph.serialize(format="json-ld")
+
+        key_root = f"catalogues/{short_path}/annotations/{uuid}"
+
+        return [
+            Messager.S3UploadAction(
+                key=key_root + ".ttl",
+                file_body=turtle,
+                mime_type="text/turtle",
+                cache_control=str(cache_control),
+                bucket=self.output_bucket,
+            ),
+            Messager.S3UploadAction(
+                key=key_root + ".jsonld",
+                file_body=jsonld,
+                mime_type="application/ld+json",
+                cache_control=str(cache_control),
+                bucket=self.output_bucket,
+            ),
+        ]
+
+
+def get_uuid_from_graph(file_contents):
+    """
+    Looks for a line containing the UUID that looks like this:
+    `owl:sameAs                  <urn:uuid:12345678-1234-1234-1234-123456789012>;`
+    """
+
+    pattern = r"owl:sameAs\s+<urn:uuid:([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})>;"  # noqa:E501
```

Comment:
This is not a reliable way to find the QA check result's UUID. There could be multiple things in the file with a UUID assigned in this way, the file could be formatted different (eg, with newlines), `owl:sameAs` might be replaced with its full URI or with a different binding, there could be a space before the `;` or it could be a comma or dot instead, etc.

This can extract the UUID from some QA results in a test file (pass the ConjunctiveGraph into this function not a string - you'll need to change Graph to ConjunctiveGraph because it contains nested named graphs):

```
from rdflib import ConjunctiveGraph, URIRef
from rdflib.namespace import RDF, OWL

# Parse the QA results. We must use a ConjunctiveGraph because the EODHQualityMeasurementDataset may be inside
# a named graph nested inside the main graph.
g = ConjunctiveGraph()
g.parse("../annotations-ingester/ontology/qa-output-1.trig")

# The entire QA run outputs are represented by a resource of type `eodhqa:EODHQualityMeasurementDataset`
# There should be only one.
eodhqa = URIRef("https://eodatahub.org.uk/api/ontologies/qa/")
measurement_dset = next(g.triples((None, RDF.type, eodhqa + "EODHQualityMeasurementDataset")))[0]

# The eodhqa:EODHQualityMeasurementDataset should always have an `owl:sameAs` triple with its UUID.
uuid_ref = next(obj for obj in list(g.objects(measurement_dset, OWL.sameAs)) if str(obj).startswith("urn:uuid:"))
uuid = str(uuid_ref)[len("urn:uuid:"):]
```

Reply:
Added